File: src/graph/dijkstra/dijkstraGraph.ts
import {IGraph} from "../IGraph";
import {Node} from "./node";
import {Edge} from "./edge";

export default class DijkstraGraph implements IGraph<string>{
    private queue:Array<Edge>;
    private cameFrom:Map<Node|null,Node|null>;
    private costSoFar:Map<Node,number>;
    constructor() {
        this.queue = [];
        this.cameFrom = new Map<Node | null, Node | null>();
        this.costSoFar = new Map<Node, number>();
    }

    explore(start: Node): void {
        this.reInit();
        this.exploring(start);
    }
    private exploring(start:Node):void{
        this.queue.push(new Edge(start,0));
        this.costSoFar.set(start,0);
        this.cameFrom.set(start,null);
        while(this.queue.length !== 0){
            this.queue.sort((a,b)=> b.cost - a.cost);
            let curEdge = <Edge>this.queue.shift();
            let curNode = curEdge.to;
            for(const neighbor of curNode.neighbors){
                let newCost = <number>this.costSoFar.get(neighbor.to) + neighbor.cost;
                if(!this.costSoFar.has(neighbor.to) || newCost < <number>this.costSoFar.get(neighbor.to)){
                    this.costSoFar.set(neighbor.to,newCost);
                    this.cameFrom.set(neighbor.to,curNode);
                    this.queue.push(neighbor);
                }
            }

        }
    }
    private reInit(){
        this.costSoFar = new Map<Node, number>();
        this.cameFrom = new Map<Node | null, Node | null>();
    }
    getPath(start:Node,target:Node):Array<string>{
        let arr = [];
        let curNode = target;
        while(curNode && curNode !== start){
            arr.push(curNode.value);
            curNode = <Node>this.cameFrom.get(curNode);
        }
        arr.push(start.value);
        arr.reverse();
        return arr;
    }
}

File: graph/dijkstra/DijkstraGraph.java
package graph.dijkstra;

import graph.Edge;
import graph.Graph;
import graph.INode;

import java.util.*;

public class DijkstraGraph<T> extends Graph<T> {
    private final Map<INode<T>,Float> costSoFar;
    private final Map<INode<T>,INode<T>> cameFrom;
    private final PriorityQueue<Edge<T>> queue;
    public DijkstraGraph(){
        super();
        this.costSoFar = new HashMap<>();
        this.cameFrom = new HashMap<>();
        this.queue = new PriorityQueue<>();
    }
    public DijkstraGraph(List<INode<T>> newList, Map<INode<T>, Float> costSoFar, Map<INode<T>,INode<T>> cameFrom, PriorityQueue<Edge<T>> queue){
        super(newList);
        this.costSoFar = costSoFar;
        this.cameFrom = cameFrom;
        this.queue = queue;
    }
    @Override
    protected void Exploring(INode<T> node) {
        this.queue.add(new Edge<T>(node,0));
        this.costSoFar.put(node,0f);
        this.cameFrom.put(node,null);
        while(!this.queue.isEmpty()){
            INode<T> curNode = this.queue.poll().to();
            for(Edge<T> neighbor:curNode.getNeighbors()){
                float newCost = this.costSoFar.get(node) + neighbor.cost();
                if(!this.costSoFar.containsKey(neighbor.to()) || newCost < this.costSoFar.get(neighbor.to())){
                    this.costSoFar.put(neighbor.to(),newCost);
                    this.cameFrom.put(neighbor.to(),curNode);
                    this.queue.add(neighbor);
                }
            }
        }
    }

    @Override
    protected void Reinitialize() {
        this.cameFrom.clear();
        this.costSoFar.clear();
        super.Reinitialize();
    }
    public List<INode<T>> getPath(INode<T> start,INode<T> target){
        if(!this.visited.isEmpty() && this.visited.get(0) == start){
            this.visited.clear();
        }
        INode<T> curNode = target;
        while(curNode != null && curNode != start){
            this.visited.add(curNode);
            curNode = this.cameFrom.get(curNode);
        }
        this.visited.add(start);
        Collections.reverse(this.visited);
        return this.visited;
    }
}